src/Compilers/Core/Portable/Serialization/StreamObjectReader.cs

```diff
@@ -0,0 +1,1020 @@
+        {
+            internal static readonly ObjectPool<List<object>> ListPool =
+                new ObjectPool<List<object>>(() => new List<object>(128), 2);
+
+            private readonly ReferenceMap _baseData;
+            private readonly List<object> _values = ListPool.Allocate();
+            private readonly int _baseDataCount;
+
+            private ReferenceMap(ObjectData data)
+            {
+                if (data != null)
+                {
+                    foreach (var value in data.Objects)
+                    {
+                        _values.Add(value);
+                    }
+                }
+            }
+
+            private ReferenceMap(ReferenceMap baseData)
+            {
+                Debug.Assert(baseData?._baseData == null, "Should be <= 1 level deep");
+                _baseData = baseData;
+                _baseDataCount = baseData?._values.Count ?? 0;
+            }
+
+            public void Dispose()
+            {
+                _values.Clear();
+                ListPool.Free(_values);
+            }
+
+            public static ReferenceMap Create(ObjectData data = null)
+            {
+                if (data != null)
+                {
+                    return new ReferenceMap(GetBaseReaderData(data));
+                }
+                else
+                {
+                    return new ReferenceMap(data);
+                }
+            }
+
+            private static readonly ConditionalWeakTable<ObjectData, ReferenceMap> s_baseDataMap
+                = new ConditionalWeakTable<ObjectData, ReferenceMap>();
+
+            private static ReferenceMap GetBaseReaderData(ObjectData data)
+            {
+                ReferenceMap baseData;
+                if (!s_baseDataMap.TryGetValue(data, out baseData))
+                {
+                    baseData = s_baseDataMap.GetValue(data, _data => new ReferenceMap(_data));
+                }
+
+                return baseData;
+            }
+
+            public int GetNextReferenceId()
+            {
+                _values.Add(null);
+                return _baseDataCount + _values.Count - 1;
+            }
+
+            public void AddValue(int referenceId, object value)
+            {
+                _values[referenceId - _baseDataCount] = value;
+            }
+
+            public object GetValue(int referenceId)
+            {
+                if (_baseData != null)
+                {
+                    if (referenceId < _baseDataCount)
+                    {
+                        return _baseData.GetValue(referenceId);
+                    }
+                    else
+                    {
+                        return _values[referenceId - _baseDataCount];
+                    }
+                }
+                else
+                {
+                    return _values[referenceId];
+                }
+            }
+        }
+
+        internal uint ReadCompressedUInt()
```

Comment:
is it vital that we support this complexity for ints/uints?  Are there cases where we're really saving out?